### PR TITLE
Upgrade python and java to 4.x.  Change docker-compose to separate out the batch load. 

### PR DESCRIPTION
**Pull request by jphaugla**
Upgrade python and java to 4.x.  Change docker-compose to separate out the batch load.  Increase other version where applicable.   

NOTE:  did not change the node.js as I ran into problems with that.  Simon Prickett will be looking at that later this week.

### CODE REVIEW

File: docs/010-application-development.md [hunk 1]
```diff
@@ -40,9 +40,10 @@ To run the application:
 
 This Docker Compose will start:
 
-1. RediSearch instance on port 6380, and import all movies, actors and create indexes
+1. RediStack container on port 6379.  can run redis-cli against this container
```
Comment (chayim) [inline docs/010-application-development.md:43]:
```
Redis Stack container on port 6379. You can run redis-cli against this container, once the ports are exposed.
```

File: docs/002-install-redisearch.md [hunk 2]
```diff
@@ -5,17 +5,17 @@ You have multiple ways to run RediSearch:
...
+* using [Docker](https://hub.docker.com/r/redis/redis-stack)
 
 Let's use Docker for now.
 
 **1.1 Open a terminal an run the following command**
 
 
 ```
-> docker run -it --rm --name redis-search-2 \
+> docker run -it --rm --name redis-stack \
```
Comment (chayim) [inline docs/002-install-redisearch.md:16]: As you're using the redis-stack container, perhaps you also want to expose redis insight?
```
-p 8081:8081
```

File: docs/010-application-development.md [hunk 3]
```diff
@@ -40,9 +40,10 @@ To run the application:
...
 1. The Java, Node and Python REST Services available on port 8085, 8086, 8087
 1. The frontend on port 8084
+1. A second RediStack container will start on port 6380 just to load the sample data to the redis stack instance running on port 6379.  This container exits once the sample data has been loaded to the 6379 container
```
Comment (chayim) [inline docs/010-application-development.md:46]: Redis Stack

File: sample-app/docker-compose.yaml [hunk 4]
```diff
@@ -1,22 +1,38 @@
-version: '3.7'
+version: '3.9'
 services:
 
-    # Start RediSearch and import data/indexes
+    # Start RediSearch 
     redisearch:
```
Comment (chayim) [inline sample-app/docker-compose.yaml:5]:
```
redistack
```

File: sample-app/docker-compose.yaml [hunk 5]
```diff
@@ -1,22 +1,38 @@
...
-        build:
-            context: ./redisearch-docker
-            dockerfile: Dockerfile
+        image:  redis/redis-stack:latest
+        container_name: redisearch
+        ports:
+            - "6379:6379"
```
Comment (chayim) [inline sample-app/docker-compose.yaml:9]:
```
- "6379:6379"
- "8001:8001"
```

File: sample-app/docker-compose.yaml [hunk 6]
```diff
@@ -1,22 +1,38 @@
...
+        ports:
+            - "6379:6379" 
+        networks:
+            - redisearch-gettingstarted
+        volumes:
+            - ./redisearch-docker/dataset:/dataset
+            - ./redis:/data`
+            
+    # Start redis-stack to run load script and exit 
+    redisearch2:
```
Comment (chayim) [inline sample-app/docker-compose.yaml:17]:
```
redis-stack2
```

File: sample-app/docker-compose.yaml [hunk 7]
```diff
@@ -1,22 +1,38 @@
...
+        image:  redis/redis-stack:latest
+        container_name: redisearch2
```
Comment (chayim) [inline sample-app/docker-compose.yaml:19]:
```
container_name: redis-stack-2
```

File: sample-app/docker-compose.yaml [hunk 8]
```diff
@@ -1,22 +1,38 @@
...
 services:
 
-    # Start RediSearch and import data/indexes
+    # Start RediSearch 
     redisearch:
-        build:
-            context: ./redisearch-docker
-            dockerfile: Dockerfile
+        image:  redis/redis-stack:latest
+        container_name: redisearch
```
Comment (chayim) [inline sample-app/docker-compose.yaml:7]:
```
container_name: redis-stack
```

File: sample-app/docker-compose.yaml [hunk 9]
```diff
@@ -1,22 +1,38 @@
...
         networks:
             - redisearch-gettingstarted
-        restart: always
-
+        volumes:
+            - ./redisearch-docker/dataset:/dataset
+        command:
+            [ "sh", "./dataset/import-data.sh" ]
+        depends_on:
+            - redisearch
```
Comment (chayim) [inline sample-app/docker-compose.yaml:29]:
```
- redis-stack
```

File: sample-app/docker-compose.yaml [hunk 10]
```diff
@@ -48,10 +65,12 @@ services:
         build:
             context: ./redisearch-python-rest
             dockerfile: Dockerfile
+        container_name: rest-python
         ports:
             - "8087:8087" 
         environment:
-            - REDIS_URL=redis://redisearch:6379
+            - REDIS_SERVER=redisearch
```
Comment (chayim) [inline sample-app/docker-compose.yaml:72]:
```
- REDIS_SERVER=redis-stack
```

File: sample-app/redisearch-docker/dataset/import-data.sh [hunk 11]
```diff
@@ -0,0 +1,6 @@
+sleep 15
+redis-cli -h redisearch -p 6379 < /dataset/import_actors.redis
```
Comment (chayim) [inline sample-app/redisearch-docker/dataset/import-data.sh:2]: Change to redis-stack as well

File: sample-app/redisearch-python-rest/Dockerfile [hunk 12]
```diff
@@ -1,4 +1,4 @@
-FROM python:3.7
+FROM python:3.9
```
Comment (chayim) [inline sample-app/redisearch-python-rest/Dockerfile:1]: Why not 3.10?

File: sample-app/redisearch-python-rest/requirements.txt [hunk 13]
```diff
@@ -1,5 +1,4 @@
-redis==3.5.3
-redisearch==1.0.0
-Flask==1.1.2
-python-dotenv==0.14.0
-Flask-Cors==3.0.8
\ No newline at end of file
+redis>=4.0
```
Comment (chayim) [inline sample-app/redisearch-python-rest/requirements.txt:1]: redis >= 4.3.3

File: sample-app/redisearch-python-rest/server.py [hunk 14]
```diff
@@ -1,44 +1,68 @@
...
-server_port = os.getenv("SERVER_PORT", "8087");
-redis_url = os.getenv("REDIS_URL", "redis://localhost:6379");
-redis_index = os.getenv("REDIS_INDEX", "idx:movie");
+from os import environ
+from redis.commands.search import reducers
+from redis.commands.search.query import NumericFilter, Query
+import redis.commands.search.aggregation as aggregations
+
+redis_password = ""
+if environ.get('REDIS_SERVER') is not None:
```
Comment (chayim) [inline sample-app/redisearch-python-rest/server.py:11]: Remove the if/else
```
redis_server = environ.get("REDIS_SERVER", "localhost")
sys.stdout.write(f"Using redis server on {redis_server}\n")
```
``

File: sample-app/redisearch-python-rest/server.py [hunk 15]
```diff
@@ -1,44 +1,68 @@
...
+    redis_server = environ.get('REDIS_SERVER')
+    print("passed in redis server is " + redis_server)
+else:
+    redis_server = 'localhost'
+    print("no passed in redis server variable ")
+
+if environ.get('REDIS_PORT') is not None:
```
Comment (chayim) [inline sample-app/redisearch-python-rest/server.py:18]: Similar approach for port

File: sample-app/redisearch-python-rest/server.py [hunk 16]
```diff
@@ -1,44 +1,68 @@
...
+    print("no passed in redis port variable ")
+
+if environ.get('SERVER_PORT') is not None:
+    server_port = int(environ.get('SERVER_PORT'))
+    print("passed in redis port is " + str(server_port))
+else:
+    server_port = 8087
+    print("no passed in redis port variable ")
+
+if environ.get('REDIS_INDEX') is not None:
```
Comment (chayim) [inline sample-app/redisearch-python-rest/server.py:32]: And index

File: sample-app/redisearch-python-rest/server.py [hunk 17]
```diff
@@ -1,44 +1,68 @@
...
+    redis_index = environ.get('REDIS_INDEX')
+    print("passed in category file location " + redis_index)
+else:
+    redis_index = "idx:movie"
+    print("no passed in redis index using idx:movie")
+
+if environ.get('REDIS_PASSWORD') is not None:
```
Comment (chayim) [inline sample-app/redisearch-python-rest/server.py:39]: And password

File: sample-app/redisearch-python-rest/server.py [hunk 18]
```diff
@@ -1,44 +1,68 @@
...
+    redis_password = environ.get('REDIS_PASSWORD')
+    print("passed in redis password is " + redis_password)
+
+# conn = redis.StrictRedis(redis_server, redis_port)
+if redis_password is not None:
+    conn = redis.StrictRedis(redis_server, redis_port, password=redis_password, charset="utf-8", decode_responses=True)
```
Comment (chayim) [inline sample-app/redisearch-python-rest/server.py:45]: StrictRedis is Redis now... We support it still - but perhaps we can just use the Redis class?

File: sample-app/redisearch-python-rest/server.py [hunk 19]
```diff
@@ -1,44 +1,68 @@
...
-    return {"status" : "Python REST Servicve is UP", "api" : "/api/1.0/search"}
+# @app.before_request
+# @app.route('/')
+# def home():
+#     return {"status": "Python REST Servicve is UP", "api": "/api/1.0/search"}
 
 
 @app.route('/api/1.0/movies/search')
 def search():
+    print("calling search")
```
Comment (chayim) [inline sample-app/redisearch-python-rest/server.py:61]: If we're going to use prints, we should approach using the built-in logger, or **loguru**

**Review by chayim — CHANGES_REQUESTED** [review]
Happy to help if you'd like

**Comment by jphaugla** [comment]
I made updates for all the comments and made a new push

File: docs/010-application-development.md [hunk 20]
```diff
@@ -40,9 +40,10 @@ To run the application:
 
 This Docker Compose will start:
 
-1. RediSearch instance on port 6380, and import all movies, actors and create indexes
+1. RediStack container on port 6379.  can run redis-cli against this container once the ports are exposed
 1. The Java, Node and Python REST Services available on port 8085, 8086, 8087
 1. The frontend on port 8084
+1. A second RediStack container will start on port 6380 just to load the sample data to the redis stack instance running on port 6379.  This container exits once the sample data has been loaded to the 6379 container
```
Comment (chayim) [inline docs/010-application-development.md:46]:
```
A second redis stack container will start on port 6380 to load the sample data to the redis stack instance running on port 6379. This container exits once the sample data has been loaded to the 6379 container
```

File: sample-app/docker-compose.yaml (hunk 4 above)
Comment (chayim) [inline sample-app/docker-compose.yaml:5]: Can we rename this container redis-stack? We should aimfor consistency.

File: sample-app/docker-compose.yaml (hunk 8 above)
Comment (chayim) [inline sample-app/docker-compose.yaml:7]: Also here.

File: sample-app/docker-compose.yaml (hunk 7 above)
Comment (chayim) [inline sample-app/docker-compose.yaml:19]: and here

File: sample-app/redisearch-docker/dataset/import-data.sh [hunk 21]
```diff
@@ -0,0 +1,6 @@
...
+redis-cli -h redisearch -p 6379 < /dataset/import_movies.redis
```
Comment (chayim) [inline sample-app/redisearch-docker/dataset/import-data.sh:3]: Due to my ask, WDY about 
```
redis-cli -h redis-stack -p 6379 < /dataset/import_movies.redis
```

File: sample-app/redisearch-docker/dataset/import-data.sh [hunk 22]
```diff
@@ -0,0 +1,6 @@
...
+redis-cli -h redisearch -p 6379 < /dataset/import_users.redis
+redis-cli -h redisearch -p 6379 < /dataset/import_create_index.redis
+exit
```
Comment (chayim) [inline sample-app/redisearch-docker/dataset/import-data.sh:6]: This is free, no need to actively call exit.

File: sample-app/redisearch-jedis-rest/README.md [hunk 23]
```diff
@@ -28,13 +28,13 @@ This command will create a new image and build the maven project into it.
 
 ```shell script
 > docker run --rm  \
-     --env "REDIS_URL=redis://host.docker.internal:6379" \
+     --env "REDIS_URL=redis://redis-stack:6379" \
```
Comment (chayim) [inline sample-app/redisearch-jedis-rest/README.md:31]: May need to be updated due to compose. In compose the current names are redisearch.

**Review by chayim — CHANGES_REQUESTED** [review]
Updated asks!

File: sample-app/redisearch-jedis-rest/README.md (hunk 23 above)
Comment (jphaugla) [reply sample-app/redisearch-jedis-rest/README.md:31]: this is already showing changed on my repository so not sure about this...

File: sample-app/redisearch-docker/dataset/import-data.sh (hunk 22 above)
Comment (jphaugla) [reply sample-app/redisearch-docker/dataset/import-data.sh:6]: removed the exit

File: docs/010-application-development.md (hunk 20 above)
Comment (jphaugla) [reply docs/010-application-development.md:46]: Not clear what you want here.  Your text seems the same as mine.   Could you just accept the TR and then make this slight change afterwards?

File: docs/010-application-development.md [hunk 24]
```diff
@@ -40,9 +40,10 @@ To run the application:
 
 This Docker Compose will start:
 
-1. RediSearch instance on port 6380, and import all movies, actors and create indexes
+1. RediStack container on port 6379.  can run redis-cli against this container once the ports are exposed
```
[the remark this answers is not in the document]
Comment (jphaugla) [reply docs/010-application-development.md:43]: I think my text is same as yours-can you just accept and make any changes after PR is accepted?

**Review by chayim — APPROVED** [review]

**Pull request merged**